File: src/core/gui/Button.ts
import { button, Icon, icon, span } from "./dom";
import "./Button.css";
import { Observable } from "../observable/Observable";
import { emitter } from "../observable";
import { Control, ControlOptions } from "./Control";
import { Property } from "../observable/property/Property";
import { WritableProperty } from "../observable/property/WritableProperty";
import { WidgetProperty } from "../observable/property/WidgetProperty";

export type ButtonOptions = ControlOptions & {
    readonly text?: string | Property<string>;
    readonly icon_left?: Icon;
    readonly icon_right?: Icon;
};

export class Button extends Control {
    private readonly _onmouseup = emitter<MouseEvent>();
    private readonly _onclick = emitter<MouseEvent>();
    private readonly _onkeydown = emitter<KeyboardEvent>();
    private readonly _text: WidgetProperty<string>;
    private readonly center_element: HTMLSpanElement;

    readonly element = button({ className: "core_Button" });
    readonly onmouseup: Observable<MouseEvent> = this._onmouseup;
    readonly onclick: Observable<MouseEvent> = this._onclick;
    readonly onkeydown: Observable<KeyboardEvent> = this._onkeydown;
    readonly text: WritableProperty<string>;

    constructor(options?: ButtonOptions) {
        super(options);

        const inner_element = span({ className: "core_Button_inner" });

        if (options?.icon_left != undefined) {
            inner_element.append(span({ className: "core_Button_left" }, icon(options.icon_left)));
        }

        this.center_element = span({ className: "core_Button_center" });
        inner_element.append(this.center_element);

        if (options?.icon_right != undefined) {
            inner_element.append(
                span({ className: "core_Button_right" }, icon(options.icon_right)),
            );
        }

        this.element.onmouseup = (e: MouseEvent) => this._onmouseup.emit({ value: e });
        this.element.onclick = (e: MouseEvent) => this._onclick.emit({ value: e });
        this.element.onkeydown = (e: KeyboardEvent) => this._onkeydown.emit({ value: e });

        this._text = new WidgetProperty<string>(this, "", this.set_text);
        this.text = this._text;

        if (typeof options?.text === "string") {
            this.text.val = options.text;
        } else if (options?.text) {
            this.text.bind_to(options.text);
        } else {
            this.text.val = "";
        }

        this.element.append(inner_element);

        this.finalize_construction();
    }

    protected set_enabled(enabled: boolean): void {
        super.set_enabled(enabled);
        this.element.disabled = !enabled;
    }

    protected set_text(text: string): void {
        this.center_element.textContent = text;
        this.center_element.hidden = text === "";
    }
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/Button.kt
package world.phantasmal.webui.widgets

import org.w3c.dom.Node
import org.w3c.dom.events.KeyboardEvent
import org.w3c.dom.events.MouseEvent
import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.nullCell
import world.phantasmal.observable.cell.trueCell
import world.phantasmal.webui.dom.Icon
import world.phantasmal.webui.dom.button
import world.phantasmal.webui.dom.icon
import world.phantasmal.webui.dom.span

open class Button(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    tooltip: Cell<String?> = nullCell(),
    private val className: String? = null,
    private val text: String? = null,
    private val textCell: Cell<String>? = null,
    private val iconLeft: Icon? = null,
    private val iconRight: Icon? = null,
    private val onMouseDown: ((MouseEvent) -> Unit)? = null,
    private val onMouseUp: ((MouseEvent) -> Unit)? = null,
    private val onClick: ((MouseEvent) -> Unit)? = null,
    private val onKeyDown: ((KeyboardEvent) -> Unit)? = null,
    private val onKeyUp: ((KeyboardEvent) -> Unit)? = null,
    private val onKeyPress: ((KeyboardEvent) -> Unit)? = null,
) : Control(visible, enabled, tooltip) {
    override fun Node.createElement() =
        button {
            className = "pw-button"

            this@Button.className?.let { classList.add(it) }

            onmousedown = onMouseDown
            onmouseup = onMouseUp
            onclick = onClick
            onkeydown = onKeyDown
            onkeyup = onKeyUp
            onkeypress = onKeyPress

            span {
                className = "pw-button-inner"

                iconLeft?.let {
                    span {
                        className = "pw-button-left"
                        icon(iconLeft)
                    }
                }

                span {
                    className = "pw-button-center"

                    if (textCell != null) {
                        observeNow(textCell) {
                            textContent = it
                            hidden = it.isEmpty()
                        }
                    } else if (!text.isNullOrEmpty()) {
                        textContent = text
                    } else {
                        hidden = true
                    }
                }

                iconRight?.let {
                    span {
                        className = "pw-button-right"
                        icon(iconRight)
                    }
                }
            }
        }

    companion object {
        init {
            @Suppress("CssUnusedSymbol", "CssUnresolvedCustomProperty")
            // language=css
            style("""
                .pw-button {
                    display: inline-flex;
                    flex-direction: row;
                    align-items: stretch;
                    align-content: stretch;
                    box-sizing: border-box;
                    height: 24px;
                    padding: 0;
                    border: var(--pw-control-border);
                    color: var(--pw-control-text-color);
                    outline: none;
                    font-size: 12px;
                    font-family: var(--pw-font-family), sans-serif;
                    overflow: hidden;
                }

                .pw-button .pw-button-inner {
                    flex-grow: 1;
                    display: inline-flex;
                    flex-direction: row;
                    align-items: center;
                    box-sizing: border-box;
                    background-color: var(--pw-control-bg-color);
                    height: 22px;
                    padding: 2px 3px;
                    border: var(--pw-control-inner-border);
                    overflow: hidden;
                }

                .pw-button:hover .pw-button-inner {
                    background-color: var(--pw-control-bg-color-hover);
                    border-color: hsl(0, 0%, 40%);
                    color: var(--pw-control-text-color-hover);
                }

                .pw-button:active .pw-button-inner {
                    background-color: hsl(0, 0%, 20%);
                    border-color: hsl(0, 0%, 30%);
                    color: hsl(0, 0%, 75%);
                }

                .pw-button:focus-within .pw-button-inner {
                    border: var(--pw-control-inner-border-focus);
                }

                .pw-button:disabled .pw-button-inner {
                    background-color: hsl(0, 0%, 15%);
                    border-color: hsl(0, 0%, 25%);
                    color: hsl(0, 0%, 55%);
                }

                .pw-button-inner > * {
                    display: inline-block;
                    margin: 0 3px;
                }

                .pw-button-center {
                    flex-grow: 1;
                    text-align: left;
                    overflow: hidden;
                    text-overflow: ellipsis;
                    white-space: nowrap;
                }

                .pw-button-left,
                .pw-button-right {
                    display: inline-flex;
                    align-content: center;
                    font-size: 11px;
                }
            """.trimIndent())
        }
    }
}
